From 34b40cbe1cd782a79d0bbb350c5dfafcaf582716 Mon Sep 17 00:00:00 2001
From: supermepsipax <supermepsipax@users.noreply.github.com>
Date: Thu, 5 Feb 2026 00:00:00 +0000
Subject: [PATCH 36/41] update v4l2 subdev state helpers and fix route access
 in ipu-isys-subdev

The v4l2_subdev_get_try_format/crop/compose helpers were removed.
Use v4l2_subdev_state_get_format/crop/compose instead (which take
state and pad, without the sd argument).

Also, v4l2_subdev_routing.routes is now __u64 (userspace pointer),
not a struct pointer. Cast it to struct v4l2_subdev_route * for
kernel-side access in ipu_isys_subdev_get_routing.
---
 drivers/media/pci/intel/ipu-isys-subdev.c | 39 ++++++++---------------
 1 file changed, 14 insertions(+), 25 deletions(-)

diff --git a/drivers/media/pci/intel/ipu-isys-subdev.c b/drivers/media/pci/intel/ipu-isys-subdev.c
index 886ceb4e658e..ca4535398713 100644
--- a/drivers/media/pci/intel/ipu-isys-subdev.c
+++ b/drivers/media/pci/intel/ipu-isys-subdev.c
@@ -158,9 +158,7 @@ struct v4l2_mbus_framefmt *__ipu_isys_get_ffmt(struct v4l2_subdev *sd,
 	if (which == V4L2_SUBDEV_FORMAT_ACTIVE)
 		return &asd->ffmt[pad][stream];
 	else
-		return v4l2_subdev_get_try_format(
-							 sd,
-							 cfg, pad);
+		return v4l2_subdev_state_get_format(cfg, pad);
 }
 
 struct v4l2_rect *__ipu_isys_get_selection(struct v4l2_subdev *sd,
@@ -180,13 +178,9 @@ struct v4l2_rect *__ipu_isys_get_selection(struct v4l2_subdev *sd,
 	} else {
 		switch (target) {
 		case V4L2_SEL_TGT_CROP:
-			return v4l2_subdev_get_try_crop(
-							       sd,
-							       cfg, pad);
+			return v4l2_subdev_state_get_crop(cfg, pad);
 		case V4L2_SEL_TGT_COMPOSE:
-			return v4l2_subdev_get_try_compose(
-								  sd,
-								  cfg, pad);
+			return v4l2_subdev_state_get_compose(cfg, pad);
 		}
 	}
 	WARN_ON(1);
@@ -629,35 +623,36 @@ int ipu_isys_subdev_get_routing(struct v4l2_subdev *sd,
 				struct v4l2_subdev_routing *route)
 {
 	struct ipu_isys_subdev *asd = to_ipu_isys_subdev(sd);
+	struct v4l2_subdev_route *routes = (struct v4l2_subdev_route *)route->routes;
 	int i, j;
 
 	for (i = 0, j = 0; i < min(asd->nstreams, route->num_routes); ++i) {
-		route->routes[j].sink_pad = asd->route[i].sink;
+		routes[j].sink_pad = asd->route[i].sink;
 		if (sd->entity.pads[asd->route[i].sink].flags &
 		    MEDIA_PAD_FL_MULTIPLEX) {
 			int source_pad = asd->route[i].source - asd->nsinks;
 
-			route->routes[j].sink_stream =
+			routes[j].sink_stream =
 			    asd->stream[asd->route[i].sink].
 			    stream_id[source_pad];
 		} else {
-			route->routes[j].sink_stream =
+			routes[j].sink_stream =
 			    asd->stream[asd->route[i].sink].stream_id[0];
 		}
 
-		route->routes[j].source_pad = asd->route[i].source;
+		routes[j].source_pad = asd->route[i].source;
 		if (sd->entity.pads[asd->route[i].source].flags &
 		    MEDIA_PAD_FL_MULTIPLEX) {
-			route->routes[j].source_stream =
+			routes[j].source_stream =
 			    asd->stream[asd->route[i].source].stream_id[asd->
 									route
 									[i].
 									sink];
 		} else {
-			route->routes[j].source_stream =
+			routes[j].source_stream =
 			    asd->stream[asd->route[i].source].stream_id[0];
 		}
-		route->routes[j++].flags = asd->route[i].flags;
+		routes[j++].flags = asd->route[i].flags;
 	}
 
 	route->num_routes = j;
@@ -879,17 +874,11 @@ int ipu_isys_subdev_open(struct v4l2_subdev *sd, struct v4l2_subdev_fh *fh)
 
 	for (i = 0; i < asd->sd.entity.num_pads; i++) {
 		struct v4l2_mbus_framefmt *try_fmt =
-			v4l2_subdev_get_try_format(
-			sd, fh->state,
-			i);
+			v4l2_subdev_state_get_format(fh->state, i);
 		struct v4l2_rect *try_crop =
-			v4l2_subdev_get_try_crop(
-			sd, fh->state,
-			i);
+			v4l2_subdev_state_get_crop(fh->state, i);
 		struct v4l2_rect *try_compose =
-			v4l2_subdev_get_try_compose(
-			sd, fh->state,
-			i);
+			v4l2_subdev_state_get_compose(fh->state, i);
 
 		*try_fmt = asd->ffmt[i][0];
 		*try_crop = asd->crop[i];
-- 
2.51.0

